File: day5_part1_2017.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

int main() {
    std::ifstream inputFile("input.txt");
    std::vector<int> offsets;
    std::string line;

    while (std::getline(inputFile, line)) {
        offsets.push_back(std::stoi(line));
    }

    int index = 0;
    int steps = 0;

    while (index >= 0 && index < offsets.size()) {
        int jump = offsets[index];
        offsets[index]++;
        index += jump;
        steps++;
    }

    std::cout << steps << std::endl;

    return 0;
}

File: day5_part1_2017.hs

import System.IO

main = do
    contents <- readFile "input.txt"
    let offsets = map read $ lines contents
    print $ solveMaze offsets 0 0

solveMaze :: [Int] -> Int -> Int -> Int
solveMaze offsets index steps
    | index < 0 || index >= length offsets = steps
    | otherwise = solveMaze newOffsets newIndex (steps + 1)
    where jump = offsets !! index
          newOffsets = incrementOffset offsets index
          newIndex = index + jump

incrementOffset :: [Int] -> Int -> [Int]
incrementOffset offsets index = 
    take index offsets ++ [offsets !! index + 1] ++ drop (index + 1) offsets
